### PR TITLE
Refactoring asserts

### PR DESCRIPTION
**Pull request by leoisl**
We have hundreds of asserts to refactor, and I think somewhere around 30-40% are done, but before proceeding I am creating this draft PR to check if you agree on the way this refactoring is being done, and to look for suggestions. This will close https://github.com/rmcolq/pandora/issues/166 and https://github.com/rmcolq/pandora/issues/169

When an `assert` is found, I try to understand if:
1. Can be safely removed; or
2. Should be kept (i.e. it is indeed an `assert`);
3. Should be refactored into a runtime check;

Most of the cases fall into case 3. One example of case 1 is when we clear a vector and check if it is empty (I think this check is not needed), e.g. https://github.com/rmcolq/pandora/commit/179b50a1632fec068cceb2010bee33df454381c9#diff-3c7cce26230366838c6688e1c8123e31418974c32bbec93107603a5d4e17e275L86-L88

Case 2 happens when I am sure there is no need for this check during runtime (i.e. it is safe for the compiler to remove the assert from the binary, it is just needed for developers/DEBUG mode). With most cases falling into case 3, most checks that were done with `assert`s are now done with `if`s and they are kept in the final binary, so most of the sanity checks are still done. I think this is a safe approach, as deciding how to refactor an `assert` is not that easy, as sometimes I am not 100% sure if the check can be safely removed.

As most of the checks that the `assert`s do are critical errors, I decided to not throw any exceptions, but straight error out. I am not sure this is the correct decision. Central to this is the `FatalError` class, which prints some fatal error messages and error out on the destructor: https://github.com/rmcolq/pandora/blob/179b50a1632fec068cceb2010bee33df454381c9/include/utils.h#L105-L121

It can be used as a stream, e.g.: https://github.com/rmcolq/pandora/blob/179b50a1632fec068cceb2010bee33df454381c9/src/estimate_parameters.cpp#L51-L54

Besides this, in each refactoring, we have a bit more documentation as the boolean condition is being named and we error out always with a meaningful message. An example is this refactoring: https://github.com/rmcolq/pandora/commit/179b50a1632fec068cceb2010bee33df454381c9#diff-d62b621c539084308a69f69ae6916ef8b6b4ff5526a9cdfd7eb941dead59bd2fL51-R54

With this draft PR, you have an idea how the refactoring is being done, and if I should change/improve something or proceed like this.

### CODE REVIEW

File: src/prg/path.cpp [hunk 1]
```diff
@@ -62,7 +62,7 @@ std::vector<LocalNodePtr> prg::Path::nodes_along_path(const LocalPRG& localPrg)
         // redudant call, return the memoized local node path
         return memoizedLocalNodePath;
     } else {
-        fatal_error("Bug on prg::Path::nodes_along_path()");
+        FatalError() << "Bug on prg::Path::nodes_along_path()";
```
Comment (mbhall88) [inline src/prg/path.cpp:65]: More descriptive error message

**Review by mbhall88 — CHANGES_REQUESTED** [review]
I am also assuming you will be updating tests according to check the errors are thrown?

**Comment by leoisl** [comment]
> I am also assuming you will be updating tests according to check the errors are thrown?

Yep, doing this now!

File: include/utils.h [hunk 2]
```diff
@@ -101,6 +102,24 @@ uint32_t pangraph_from_read_file(const std::string&, std::shared_ptr<pangenome::
 void infer_most_likely_prg_path_for_pannode(
     const std::vector<std::shared_ptr<LocalPRG>>&, PanNode*, uint32_t, float);
 
+class FatalError {
+public:
+    FatalError(){
+        BOOST_LOG_TRIVIAL(error) << std::endl << std::endl << "[FATAL ERROR] ";
```
[the remark this answers is not in the document]
Comment (leoisl) [reply include/utils.h]: Ok, changed!

File: src/kmergraph.cpp [hunk 3]
```diff
@@ -105,12 +97,15 @@ KmerNodePtr KmerGraph::add_node(const prg::Path& p)
     KmerNodePtr n(std::make_shared<KmerNode>(nodes.size(), p)); // create the node
     nodes.push_back(n); // add it to nodes
     sorted_nodes.insert(n);
-    assert(k == 0 or p.length() == 0 or p.length() == k);
+
+    bool path_is_valid = k == 0 or p.length() == 0 or p.length() == k;
+    if (!path_is_valid) {
+        FatalError() << "In KmerGraph::add_node(), the node path is not valid (k is " << k
```
[the remark this answers is not in the document]
Comment (leoisl) [reply src/kmergraph.cpp:103]: Yeah, this is a bit subjective. On one end, describing errors in plain english is good for the user to understand it, but on another end, some of the errors are just implementation details, and the description is not useful for the user, as they won't be able to understand them. For many errors, the description is targeted to developers than to the users. Some errors I think are not necessary, but I am not removing any as of now, to keep consistency with the previous versions. But I also think I did not do this error logging it in the proper way, and thus this has been confusing for you to review. So I guess the best to do is to provide a plain english message for the user, but also a decent error logging for developers (will be using [Boost Stacktrace](https://www.boost.org/doc/libs/1_75_0/doc/html/stacktrace.html)) for that. I know that we can always ask the user for the input, and debug the error in debug executions, but many cases just a look at the stack trace might be enough to know already the issue!

File: src/kmergraphwithcoverage.cpp [hunk 4]
```diff
@@ -15,22 +15,35 @@ using namespace prg;
 
 void KmerGraphWithCoverage::set_exp_depth_covg(const uint32_t edp)
 {
-    assert(edp > 0);
+    bool exp_depth_covg_parameter_is_valid = edp > 0;
+    if (!exp_depth_covg_parameter_is_valid) {
```
[the remark this answers is not in the document]
Comment (leoisl) [reply src/kmergraphwithcoverage.cpp:19]: In almost all cases, a bool explaining the condition is good for documentation purposes. I will leave this one also, as there is nothing we actually lose in terms of code and documentation. If this condition gets more complicated in the future, we already have this structure... and I am changing all conditions to this standard, so will keep following this standard. Please reopen if you feel we should not use the bool.

File: src/kmergraphwithcoverage.cpp [hunk 5]
```diff
@@ -455,12 +521,29 @@ void KmerGraphWithCoverage::load(const std::string& filepath)
...
-                    or assert_msg(stoi(split_line[3]) << ">=" << innode_counts.size()));
+
+                bool line_is_consistent = split_line.size() >= 5;
+                if (!line_is_consistent) {
+                    FatalError() << "In KmerGraphWithCoverage::load(), line \"" << line << "\" "
+                                 << "is inconsistent";
+                }
+
+                int from_node = stoi(split_line[1]);
+                int to_node = stoi(split_line[3]);
```
[the remark this answers is not in the document]
Comment (leoisl) [reply src/kmergraphwithcoverage.cpp:532]: Will add the `const`, I dislike `auto` as the type is good for documentation purposes for me. But that is more like a personal opinions (generally, when I see `auto`, I have to navigate into the function and see the return type to infer the type of the variable). If you feel strong about using `auto`, please reopen.

File: src/kmergraph.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (leoisl) [reply src/kmergraph.cpp:103]: [Boost Stacktrace](https://www.boost.org/doc/libs/1_75_0/doc/html/stacktrace.html) does not work on our project :( ... Issue is that it was introduced in `Boost 1.65.0`, but we are stuck at `Boost 1.62.0` because of `GATB` (`GATB` introduces a local copy of `Boost 1.62.0`, and it seems that linking any different boost versions give us linker errors... I also searched how to use two different `Boost` versions in a single project, but this can result in runtime errors, so it is not advisable). I ended up using [backward-cpp](https://github.com/bombela/backward-cpp/) for this, and now we provide a plain english error message and a stack trace that will probably help us debug the issue with the user not even sending us the input, and etc. Do you think it is fine to keep it like this @mbhall88 ? An example in a test case that fails with a bug is below

```
[FATAL ERROR]: Negative binomial parameters are invalid (mean is 0.000000, variance is 1.000000)
Aborting...

DEBUG info (for developers, provide this if opening an issue):
Stack trace (most recent call last):
#18   Object "", at 0xffffffffffffffff, in 
#17   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc3a45fd, in _start
#16   Object "/usr/lib/x86_64-linux-gnu/libc-2.31.so", at 0x7f9be93070b2, in __libc_start_main
      Source "../csu/libc-start.c", line 308, in __libc_start_main [0x7f9be93070b2]
#15   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc93ac06, in main
      Source "/home/leandro/git/pandora/test/main.cpp", line 10, in main [0x557abc93ac06]
          7:     core->set_logging_enabled(false);
          8: 
          9:     ::testing::InitGoogleTest(&argc, argv);
      >  10:     return RUN_ALL_TESTS();
         11: }
#14   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc93ad8e, in RUN_ALL_TESTS()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Install/include/gtest/gtest.h", line 2473, in RUN_ALL_TESTS [0x557abc93ad8e]
       2470: int RUN_ALL_TESTS() GTEST_MUST_USE_RESULT_;
       2471: 
       2472: inline int RUN_ALL_TESTS() {
      >2473:   return ::testing::UnitTest::GetInstance()->Run();
       2474: }
       2475: 
       2476: GTEST_DISABLE_MSC_WARNINGS_POP_()  //  4251
#13   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcca6d7f, in testing::UnitTest::Run()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 4925, in Run [0x557abcca6d7f]
#12   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abccbfc50, in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2469, in HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> [0x557abccbfc50]
#11   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abccc63a9, in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2433, in HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> [0x557abccc63a9]
#10   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcca853f, in testing::internal::UnitTestImpl::RunAllTests()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 5338, in RunAllTests [0x557abcca853f]
#9    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcc9c483, in testing::TestSuite::Run()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2816, in Run [0x557abcc9c483]
#8    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcc9bd48, in testing::TestInfo::Run()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2684, in Run [0x557abcc9bd48]
#7    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcc9b35d, in testing::Test::Run()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2508, in Run [0x557abcc9b35d]
#6    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abccbec42, in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2469, in HandleExceptionsInMethodIfSupported<testing::Test, void> [0x557abccbec42]
#5    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abccc4f80, in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2433, in HandleSehExceptionsInMethodIfSupported<testing::Test, void> [0x557abccc4f80]
#4    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc7c7316, in EstimateParameters_FitNegativeBinomial_MeanZero_Death_Test::TestBody()
      Source "/home/leandro/git/pandora/test/estimate_parameters_test.cpp", line 68, in TestBody [0x557abc7c7316]
         65: {
         66:     double mean = 0, variance = 1;
         67:     float p, r;
      >  68:     ASSERT_EXCEPTION(fit_negative_binomial(mean, variance, p, r), runtime_error,
         69:         "Negative binomial parameters are invalid");
         70: }
#3    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc4cfdc7, in fit_negative_binomial(double&, double&, float&, float&)
      Source "/home/leandro/git/pandora/src/estimate_parameters.cpp", line 51, in fit_negative_binomial [0x557abc4cfdc7]
         48: {
         49:     bool negative_binomial_parameters_are_ok = mean > 0 and variance > 0 and mean < variance;
         50:     if (!negative_binomial_parameters_are_ok) {
      >  51:         fatal_error("Negative binomial parameters are invalid (mean is ", mean,
         52:             ", variance is ", variance, ")");
         53:     }
         54:     p = mean / variance;
#2    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc4d5428, in void fatal_error<char [51], double, char [15], double, char [2]>(char const (&) [51], double const&, char const (&) [15], double const&, char const (&) [2])
      Source "/home/leandro/git/pandora/include/fatal_error.h", line 36, in fatal_error<char [51], double, char [15], double, char [2]> [0x557abc4d5428]
         33: template<typename... Args>
         34: void fatal_error(Args const&... args) {
         35:     backward::StackTrace stacktrace;
      >  36:     stacktrace.load_here(32);
         37:     backward::Printer stacktrace_printer;
         38:     stacktrace_printer.address = true;
         39:     stacktrace_printer.object = true;
#1    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc3a58bd, in backward::StackTraceImpl<backward::system_tag::linux_tag>::load_here(unsigned long, void*, void*)
      Source "/home/leandro/git/pandora/include/backward.hpp", line 859, in load_here [0x557abc3a58bd]
        856:             return 0;
        857:         }
        858:         _stacktrace.resize(depth);
      > 859:         size_t trace_cnt = details::unwind(callback(*this), depth);
        860:         _stacktrace.resize(trace_cnt);
        861:         skip_n_firsts(0);
        862:         return size();
#0    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc3aae4c, in unsigned long backward::details::unwind<backward::StackTraceImpl<backward::system_tag::linux_tag>::callback>(backward::StackTraceImpl<backward::system_tag::linux_tag>::callback, unsigned long)
      Source "/home/leandro/git/pandora/include/backward.hpp", line 841, in unwind<backward::StackTraceImpl<backward::system_tag::linux_tag>::callback> [0x557abc3aae4c]
        839: template <typename F> size_t unwind(F f, size_t depth) {
        840:     Unwinder<F> unwinder;
      > 841:     return unwinder(f, depth);
        842: }
        843: 
        844: } // namespace details
```

[the remark this answers is not in the document]
Comment (mbhall88) [reply src/kmergraph.cpp:103]: Great workaround! Looks good to me. Do they have a way to control the verbosity of the stacktrace? We could potentially link that to the `-v` flag? I don't mind leaving it as is though...

File: src/fastaq.cpp [hunk 6]
```diff
@@ -51,7 +49,6 @@ void Fastaq::add_entry(const std::string& name, const std::string& sequence,
     const std::vector<uint32_t>& covgs, const uint_least16_t global_covg,
     const std::string header)
 {
-
     assert(name != "");
     assert(covgs.size() == sequence.length());
```
[the remark this answers is not in the document]
Comment (leoisl) [reply src/fastaq.cpp:53]: Yeah, I went through these ones, and actually we (as devs) are composing the Fasta entries to be added to the Fasta file. I previously kept these as `assert`s as I don't think these erroneous conditions can happen in runtime if not due to a developer error. But this is the case for current code, in the future, we might use it in another way, so you are right, throwing an exception is the safer and better option

File: src/kmergraph.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (leoisl) [reply src/kmergraph.cpp]: As per https://github.com/bombela/backward-cpp/#printer , we can remove the `Object` line and the address of the source location (although I think this second one is not advisable...)

File: src/kmergraph.cpp [hunk 7]
```diff
@@ -285,7 +301,13 @@ void KmerGraph::load(const fs::path& filepath)
...
             if (line[0] == 'S') {
                 split_line = split(line, "\t");
-                assert(split_line.size() >= 4);
+
+                bool line_is_consistent = split_line.size() >= 4;
+                if (!line_is_consistent) {
+                    FatalError() << "In KmerGraph::load(), line \"" << line << "\" "
+                                 << "is inconsistent";
+                }
+
```
[the remark this answers is not in the document]
Comment (leoisl) [reply src/kmergraph.cpp]: There should have, but I guess adding a GFA library is out of the scope of this PR

File: src/kmergraph.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbhall88) [reply src/kmergraph.cpp]: No worries. Leave it as is :+1:

File: src/kmergraph.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbhall88) [reply src/kmergraph.cpp]: Yeah, agreed.

**Comment by mbhall88** [comment]
Could you also make sure you update the changelog accordingly? Might need to also add the stuff from #256 as that PR didn't update the changelog

**Comment by leoisl** [comment]
This is mostly done, I am just testing with a real dataset, and also to provide new benchmarks for the paper review

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,6 +9,15 @@ project adheres to
...
 
+Some improvements to the build process:
+* We now use the [Hunter](https://github.com/cpp-pm/hunter) package manager to remove the requirement of having ZLIB and
+Boost system-wide installations;
+* We now provide a portable precompiled binary as another option to run `pandora` easily;
+* We refactored asserts into exceptions, and now `pandora` can be compiled correctly in the `Release` mode, and the build
+process will thus be able to create a more optimized binary;
+* `pandora` can now provide a meaningful stack trace in case of errors, to facilitate debugging
+  (need to pass flag `-DPRINT_STACKTRACE` to `CMake`);
+
```
Comment (mbhall88) [inline CHANGELOG.md:20]: Very minor (fussy) request: Could you please use the "Types of changes" headings, as in https://keepachangelog.com/en/1.0.0/ . i.e.e "Added", "Changed", "Removed" etc.

File: README.md [hunk 9]
```diff
@@ -83,13 +83,13 @@ can be found [here](https://gcc.gnu.org/onlinedocs/gfortran/OpenMP.html).
...
-  wget "https://www.dropbox.com/s/74ptrnk4k5qcc6o/pandora-linux-precompiled_v0.8.0_beta?dl=1" -O pandora-linux-precompiled_v0.8.0_beta
+  wget "https://www.dropbox.com/s/ltq2gti9t6wav1j/pandora-linux-precompiled_v0.8.1_beta?dl=1" -O pandora-linux-precompiled_v0.8.1_beta
   ```
   * **TODO: updated to a github link when we make the release;**
 * **Running**:
 ```
-chmod +x pandora-linux-precompiled_v0.8.0_beta
-./pandora-linux-precompiled_v0.8.0_beta -h
+chmod +x pandora-linux-precompiled_v0.8.1_beta
+./pandora-linux-precompiled_v0.8.1_beta -h
```
Comment (mbhall88) [inline README.md:92]: This is unnecessary as we haven't even released v0.8.0 yet

Comment (leoisl) [reply README.md:92]: That is true, v0.8.0 is the one that has the build process changes and provide the static binary, v0.8.1 is the one with asserts refactored. Will fix this.

Comment (mbhall88) [reply README.md:92]: Well, given none of those things have been officially released it is all v0.8.0...

Comment (leoisl) [reply README.md:92]: Both are fine for me, was my mistake pushing my private versioning. I guess will be able to merge this next week, will have the fix

Comment (leoisl) [reply README.md]: We are now at version [0.8.0-alpha](https://github.com/rmcolq/pandora/releases/tag/v0.8.0-alpha)

**Comment by leoisl** [comment]
This PR is ready for review, although there are a few TODO items, I think. Its tip is the [v0.8.0-alpha prerelease](https://github.com/rmcolq/pandora/releases/tag/v0.8.0-alpha).

TODO:
* Apply clang format (I did not do it because then it would be even more complicated to review - there are several commits previous to this PR that did not have clang format applied);
* There are a few more tests to check the exceptions are thrown, but they do not come close to cover all exceptions. There are ~170 points where exceptions can be thrown, so this would mean also adding at least 170 test cases (some of these are already in the code base). We can of course do this, just tell me, but it might be fine to just leave as it is and fix issues as we find them, and progressively add more test cases;
* I think some of these checks could be removed, but I guess it is better to keep them just as sanity check. This however costs performance, mainly when the compiler predicts the outcome of the condition wrongly. 99.9% of these checks result in no error, and if an error is detected, then it is fine for the compiler to miss the prediction as i) it is a rare situation; ii) the execution will end with an error anyway, so performance does not matter. Unfortunately there is [no standard approach to do this in C++](https://stackoverflow.com/questions/30130930/is-there-a-compiler-hint-for-gcc-to-force-branch-prediction-to-always-go-a-certa) (only on C++20 or if we assume `GCC` is used...). Should we think about forcing branch prediction to go towards always the non-exception path or do you think this is a not-important optimisation?

@mbhall88 I reread and applied your previous suggestions to all changes in this PR. Sorry if I missed anything.

Changelog (also includes changes from https://github.com/rmcolq/pandora/pull/256):

Improvements to the build process and performance.

### Added
- We now provide a script to build a portable precompiled binary as another option to run `pandora` easily.
  The portable binary is now provided with the release;
- `pandora` can now provide a meaningful stack trace in case of errors, to facilitate debugging
  (need to pass flag `-DPRINT_STACKTRACE` to `CMake`). Due to this, we now add debug symbols (`-g` flag)
  to every `pandora` build type, but this [does not impact performance](https://stackoverflow.com/a/39223245).
  The precompiled binary has this enabled.
  
### Changed
- We now use the [Hunter](https://github.com/cpp-pm/hunter) package manager, removing the requirement of having `ZLIB` and
  `Boost` system-wide installations;
- `GATB` is now a git submodule instead of an external project downloaded and compiled during compilation time.
  This means that when git cloning `pandora`, `cgranges` and `GATB` are also downloaded/cloned, and when preparing
  the build (running `cmake`), `Hunter` downloads and installs `Boost`, `GTest` and `ZLIB`.
  Thus we still need internet connection to prepare the build (running `cmake`) but not for compiling (running `make`).
- We now use a GATB fork that accepts a `ZLIB` custom installation;
- Refactored all thirdparty libraries (`cgranges`, `GATB`, `backward`, `CLI11`, `inthash`) into their own directory `thirdparty`.

### Fixed
- Refactored asserts into exceptions, and now `pandora` can be compiled correctly in the `Release` mode. 
  The build process is thus able to create a more optimized binary, resulting in improved performance.

File: README.md [hunk 10]
```diff
@@ -21,6 +22,12 @@
 - [Usage](#usage)
 
 
+## News
+
+* A new pre-release version is out: [v0.8.0-alpha](https://github.com/rmcolq/pandora/releases/tag/v0.8.0-alpha).
```
Comment (mbhall88) [inline README.md:27]: I don't think this news statement is necessary

File: .gitignore [hunk 11]
```diff
@@ -117,3 +117,4 @@ example/pandora_workflow
 build_portable_executable
 
 pandora-linux-precompiled
+/cmake-build-release/
```
Comment (mbhall88) [inline .gitignore:120]: Maybe add a similar thing to the `.dockerignore` file

**Review by mbhall88 — APPROVED** [review]
Thanks for this Leandro! Feel free to do the two comments and then merge. Once you've merged I can create a PR for the cluster size fix
